### PR TITLE
[confmap] Add strict type validation under a feature gate

### PR DESCRIPTION
**Pull request by mx-psi**
#### Description

- Add `confmap.strictlyTypedInput` feature gate that introduces stricter type checks when resolving configuration
- Make `confmap.NewRetrievedFromYAML` function public so that external providers have consistent behavior when resolving YAML
- Adds `confmap.Retrieved.AsString` method to retrieve string representation for retrieved values

#### Link to tracking issue

Relates to #9854, updates #8565, #9532

### CODE REVIEW

File: confmap/expand.go [hunk 1]
```diff
@@ -111,22 +113,43 @@ func (mr *Resolver) findAndExpandURI(ctx context.Context, input string) (any, bo
...
 	}
 	expanded, changed, err := mr.expandURI(ctx, uri)
 	if err != nil {
 		return input, false, err
 	}
-	repl, err := toString(expanded)
+
+	var repl string
+	if internal.StrictlyTypedInputGate.IsEnabled() {
+		repl, err = expanded.AsString()
```
Comment (TylerHelmuth) [inline confmap/expand.go:135]: Can we add a test for this scenario?

File: confmap/provider.go [hunk 2]
```diff
@@ -116,6 +122,32 @@ func WithRetrievedClose(closeFunc CloseFunc) RetrievedOption {
...
+	var rawConf any
+	if err := yaml.Unmarshal(yamlBytes, &rawConf); err != nil {
+		return nil, err
+	}
+
+	switch v := rawConf.(type) {
+	case string:
+		opts = append(opts, withStringRepresentation(v))
+	case int, int32, int64, float32, float64, bool:
+		opts = append(opts, withStringRepresentation(string(yamlBytes)))
```
Comment (TylerHelmuth) [inline confmap/provider.go:145]: Are there test cases we can add for this?

**Review by TylerHelmuth — COMMENTED** [review]
Couple question about tests, but otherwise the e2e tests are making me feel comfortable.  the `changed` logic in expand is confusing to me, I am going to see if I can make a quick PR to remove it.

**Comment by TylerHelmuth** [comment]
Created https://github.com/open-telemetry/opentelemetry-collector/pull/10403 to try and simplify some stuff before this PR

File: confmap/provider.go (hunk 2 above)
Comment (mx-psi) [reply confmap/provider.go:145]: The end to end tests with `TargetFieldInlineString` test this. What kind of tests would you like to see?

File: confmap/expand.go (hunk 1 above)
Comment (mx-psi) [reply confmap/expand.go:135]: The end to end tests with `TargetFieldInlineString` test this. What kind of tests would you like to see?

File: confmap/provider.go (hunk 2 above)
Comment (mx-psi) [reply confmap/provider.go:145]: Ah, sorry, I think I got this wrong. You want to test this with something that does not have one of those types, right?

File: confmap/expand.go (hunk 1 above)
Comment (TylerHelmuth) [reply confmap/expand.go:135]: If the e2e test handles this we're good

File: confmap/provider.go (hunk 2 above)
Comment (TylerHelmuth) [reply confmap/provider.go:145]: If the e2e test covers this I'm good.

**Review by TylerHelmuth — APPROVED** [review]

File: confmap/expand.go [hunk 3]
```diff
@@ -121,16 +128,27 @@ func (mr *Resolver) findAndExpandURI(ctx context.Context, input string) (any, bo
...
 	return strings.ReplaceAll(input, uri, repl), true, err
 }
 
 // toString attempts to convert input to a string.
-func toString(input any) (string, error) {
+func toString(ret *Retrieved) (string, error) {
 	// This list must be kept in sync with checkRawConfType.
+	input, err := ret.AsRaw()
+	if err != nil {
+		return "", err
```
Comment (songy23) [inline confmap/expand.go:149]: Is this covered by e2e tests as well?

File: confmap/internal/featuregate.go [hunk 4]
```diff
@@ -0,0 +1,14 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package internal // import "go.opentelemetry.io/collector/confmap/internal"
+
+import "go.opentelemetry.io/collector/featuregate"
+
+const StrictlyTypedInputID = "confmap.strictlyTypedInput"
+
+var StrictlyTypedInputGate = featuregate.GlobalRegistry().MustRegister(StrictlyTypedInputID,
```
Comment (songy23) [inline confmap/internal/featuregate.go:10]: nit: curious why this needs to be in internal? There may be contrib packages that want to test against this feature gate

**Review by songy23 — COMMENTED** [review]
Some nits, LGTM overall

File: confmap/internal/featuregate.go (hunk 4 above)
Comment (mx-psi) [reply confmap/internal/featuregate.go:10]: No strict need for this to be internal, but I would like to wait until someone asks for it to make it public

File: confmap/expand.go (hunk 3 above)
Comment (mx-psi) [reply confmap/expand.go:149]: There is no way to hit this path today https://github.com/open-telemetry/opentelemetry-collector/blob/48af1b8ea63a3393e468368beed47ce3668e5c73/confmap/provider.go#L152

**Comment by mx-psi** [comment]
I am going to merge so this lands on the next release, we can discuss making it public once a component requests it

**Pull request merged**